[CSS] Support :visited as the scoping root selectors

mdubet opened the pull request:
#### d6d238642bbc3cc3fd28823eb12552f9bf7a73d8
<pre>
[CSS] Support :visited as the scoping root selectors
<a href="https://bugs.webkit.org/show_bug.cgi?id=296947">https://bugs.webkit.org/show_bug.cgi?id=296947</a>
<a href="https://rdar.apple.com/157588890">rdar://157588890</a>

Reviewed by Antti Koivisto.

We statically take the scoping root selectors into account
to determine if we are potentially resolving visited/link style.

We track wheter we have match visited during scoping root matching, and
pass this result to normal style rule matching.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited.html:

Add slightly more complex case when the scoping root is a link, but the matching is done with class selector.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::hasExplicitPseudoClassScope const):
(WebCore::CSSSelector::hasScope const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
(WebCore::SelectorChecker::determineLinkMatchType):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
(WebCore::Style::ElementRuleCollector::ruleMatches):
* Source/WebCore/style/RuleData.cpp:
(WebCore::Style::RuleData::RuleData):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::RuleData):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRuleWithSelectorList):

Canonical link: <a href="https://commits.webkit.org/299560@main">https://commits.webkit.org/299560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38c43ccaad913ec40c867b77f3c21499e3960ef3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119410 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125662 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71481 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/490ec39b-0826-4fd7-bc8c-dd31d91bf015) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47678 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90734 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2a20d8b4-cd57-46dd-ab3b-4d14b46a85d3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31762 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107064 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71212 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7656f5aa-d050-483e-9c06-6d86dab54900) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30797 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25173 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69313 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101217 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25364 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128655 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46328 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35064 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99310 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46693 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99118 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25188 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44551 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22558 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46191 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51891 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45656 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49006 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47343 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1145,9 +1145,20 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, LocalContext& c
+                return false;
+
+            if (isLink)
+                context.visitedMatchType = VisitedMatchType::Disabled;
```
what does this do?

mdubet replied:
The goal was to track whether we have match a link already (to disable visited matching). The current patch now uses a boolean return value.

anttijk commented on Source/WebCore/style/RuleSetBuilder.cpp:
```diff
@@ -292,7 +292,24 @@ void RuleSetBuilder::addStyleRuleWithSelectorList(const CSSSelectorList& selecto
+
+        RuleData ruleData(rule, selectorIndex, selectorListIndex, m_ruleSet->ruleCount(), m_isStartingStyle, scopingRootLinkMatchType);
         m_mediaQueryCollector.addRuleIfNeeded(ruleData);
         m_ruleSet->addRule(WTFMove(ruleData), m_currentCascadeLayerIdentifier, m_currentContainerQueryIdentifier, m_currentScopeIdentifier);
```
This is all so very messy. Surely there is a better factoring? Can you for example move the link match determination from constructor to RuleSet::addRule where all the relevant data should be available.

mdubet replied:
Refactored ✔️

anttijk commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -160,6 +160,28 @@ void RuleSet::addRule(RuleData&& ruleData, CascadeLayerIdentifier cascadeLayerId
+        };
+        return SelectorChecker::determineLinkMatchType(ruleData.selector(), scopingRootLinkMatchType());
+    };
+    ruleData.setLinkMatchType(computeLinkMatchType());
```
Some more of the stuff above could probably still move to SelectorChecker::determineLinkMatchType (by passing the scopeRule)

mdubet committed:
[CSS] Support :visited as the scoping root selectors
https://bugs.webkit.org/show_bug.cgi?id=296947
rdar://157588890

Reviewed by Antti Koivisto.

We statically take the scoping root selectors into account
to determine if we are potentially resolving visited/link style.

We track wheter we have match visited during scoping root matching, and
pass this result to normal style rule matching.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-visited.html:

Add slightly more complex case when the scoping root is a link, but the matching is done with class selector.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::hasExplicitPseudoClassScope const):
(WebCore::CSSSelector::hasScope const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
(WebCore::SelectorChecker::determineLinkMatchType):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
(WebCore::Style::ElementRuleCollector::scopeRulesMatch):
* Source/WebCore/style/ElementRuleCollector.h:
(WebCore::Style::ElementRuleCollector::ruleMatches):
* Source/WebCore/style/RuleData.cpp:
(WebCore::Style::RuleData::RuleData):
* Source/WebCore/style/RuleData.h:
(WebCore::Style::RuleData::RuleData):
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addStyleRuleWithSelectorList):

Canonical link: https://commits.webkit.org/299560@main
